### PR TITLE
FileServer implementation.

### LINKED ISSUE #187: Simple file server (no dispatcher, no interceptors)
Applications frequently serve static files as a response (`.js`, `.css`, images, etc) and, for these, developers  implement unexported function to detect and set the `Content-Type`, based on the files' extension, omitting the `charset` parameter often. We might want to do this properly in the `safehttp.DefaultDispatcher`.

### PR DESCRIPTION
**Pull request by kele**
In our threat model we assume that the directory passed to the
`FileServer` contains exclusively fully-trusted content.

- a `FileServerResponse` is generated when the underlying `http.FileServer`
calls `WriteHeader(200)`; the response is written as any other safe
response (e.g. goes through interceptors)
- the `DefaultDispatcher` sets the Content-Type according to what
`FileServerResponse` sets, but doesn't actually write the content due to
the nature of `http.FileServer` (many direct calls to
`http.ResponseWriter.Write([]byte)`)
- all errors coming from `http.FileServer` are converted to
`safehttp.StatusNotFound` in order to prevent information leaks

Run `go mod tidy` too.

Fixes #187.

### CODE REVIEW

**Comment by kele** [comment]
@googlebot I fixed it.

File: safehttp/fileserver.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+import (
+	"errors"
+	"net/http"
+)
+
+// FileServer returns a handler that serves HTTP requests with the contents of
+// the file system rooted at root.
+func FileServer(root string) Handler {
+	fileServer := http.FileServer(http.Dir(root))
```
Comment (mtrzos) [inline safehttp/fileserver.go:25]: Just out of curiosity - the default http.FileServer is safe from path traversal?

File: safehttp/fileserver.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+	return HandlerFunc(func(rw ResponseWriter, req *IncomingRequest) Result {
+		fileServerRW := &fileServerResponseWriter{flight: rw.(*flight), header: http.Header{}}
+		fileServer.ServeHTTP(fileServerRW, req.req)
+		return fileServerRW.result
+	})
+}
+
+type fileServerResponseWriter struct {
+	flight *flight
```
Comment (mtrzos) [inline safehttp/fileserver.go:35]: does that mean response in-flight? Is this some broader known convention in Go?

Comment (kele) [reply safehttp/fileserver.go:35]: It's the part of the framework that needs to be run per response in-flight, yes. It's not a broadly known convention in Go.

File: safehttp/fileserver.go (hunk 1 above)
Comment (kele) [reply safehttp/fileserver.go:25]: AFAIK yes, but we want extra tests to confirm: https://github.com/google/go-safeweb/issues/233.

File: safehttp/fileserver.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	return fileServerRW.header
+}
+
+func (fileServerRW *fileServerResponseWriter) Write(b []byte) (int, error) {
+	if !fileServerRW.committed {
+		fileServerRW.WriteHeader(int(StatusOK))
+	}
+
+	if fileServerRW.errored {
+		// Let the framework handle the error
```
Comment (empijei) [inline safehttp/fileserver.go:62]: Ultranit
```suggestion
		// Let the framework handle the error.
```

File: safehttp/fileserver.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+		return 0, errors.New("discarded")
+	}
+	return fileServerRW.flight.rw.Write(b)
+}
+
+func (fileServerRW *fileServerResponseWriter) WriteHeader(statusCode int) {
```
Comment (empijei) [inline safehttp/fileserver.go:68]: fileServerRW is a bit long as a name for a receiver. Consider, maybe, fsrw?

File: safehttp/fileserver.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+	// private, to not allow modifications
+	contentType string
+}
+
+// ContentType is the Content-Type of the response.
+func (resp FileServerResponse) ContentType() string {
+	return resp.contentType
+}
+
+func contentType(h http.Header) string {
```
Comment (empijei) [inline safehttp/fileserver.go:121]: This is a very generic name for something that is visible by the entire package.

Maybe consider defining this inside the only function that uses it or giving it a more specific name?

**Review by mtrzos — APPROVED** [review]

File: safehttp/fileserver.go (hunk 5 above)
Comment (kele) [reply safehttp/fileserver.go:121]: Removed altogether.

**Pull request merged**